src/tests/ftest/util/apricot/apricot/test.py

```diff
@@ -0,0 +1,180 @@
+    '''
+    Run tests without DAOS servers
+
+    :avocado: recursive
+    '''
+    def setUp(self):
+        super(TestWithoutServers, self).setUp()
+        # get paths from the build_vars generated by build
+        with open('../../../.build_vars.json') as build_vars:
+            build_paths = json.load(build_vars)
+        self.basepath = os.path.normpath(build_paths['PREFIX']  + "/../")
+        self.prefix = build_paths['PREFIX']
+        self.tmp = self.prefix + '/tmp'
+        self.daos_test = self.basepath + '/install/bin/daos_test'
+        self.orterun = self.basepath + '/install/bin/orterun'
+        self.daosctl = self.basepath + '/install/bin/daosctl'
+
+        # setup fault injection, this MUST be before API setup
+        fault_list = self.params.get("fault_list", '/run/faults/*/')
+        if fault_list:
+            # not using workdir because the huge path was messing up
+            # orterun or something, could re-evaluate this later
+            tmp = os.path.join(self.basepath, 'install', 'tmp')
+            self.fault_file = fault_config_utils.write_fault_file(tmp,
+                                                                  fault_list,
+                                                                  None)
+            os.environ["D_FI_CONFIG"] = self.fault_file
+
+        self.context = DaosContext(self.prefix + '/lib/')
+        self.d_log = DaosLog(self.context)
+
+    def tearDown(self):
+        super(TestWithoutServers, self).tearDown()
+
+        if self.fault_file:
+            os.remove(self.fault_file)
+
+class TestWithServers(TestWithoutServers):
+    '''
+    Run tests with DAOS servers
+
+    :avocado: recursive
+    '''
+    def __init__(self, *args, **kwargs):
+        super(TestWithServers, self).__init__(*args, **kwargs)
+
+        self.agent_sessions = None
+
+    def setUp(self):
+        super(TestWithServers, self).setUp()
+
+        self.server_group = self.params.get("server_group", '/server/',
+                                            'daos_server')
+        self.hostlist = self.params.get("test_machines", '/run/hosts/*')
+
+        self.hostfile = write_host_file.write_host_file(self.hostlist,
+                                                        self.workdir)
+
+        self.agent_sessions = AgentUtils.run_agent(self.basepath,
+                                                   self.hostlist)
+        server_utils.run_server(self.hostfile, self.server_group, self.basepath)
+
+    def tearDown(self):
+
+        try:
+            if self.agent_sessions:
+                AgentUtils.stop_agent(self.hostlist, self.agent_sessions)
+        finally:
+            server_utils.stop_server(hosts=self.hostlist)
+
+        super(TestWithServers, self).tearDown()
+
+class TestWithClients(TestWithServers):
+    '''
+    Run tests with DAOS servers
+
+    :avocado: recursive
+    '''
+
+    # TODO: use hostlist_servers and hostlist_clients instead of just hostlist
```

Comment:
ahh, okay.  I don't think you can leave as a todo since tests with both exist today.

Reply:
Right, but none of those are in #377 so the implementation of this class can be deferred until tests that use clients are ported to apricot.